usr/lib/common/mech_aes.c

```diff
@@ -3636,6 +3992,152 @@ CK_RV ckm_aes_key_gen(STDLL_TokData_t *tokdata, TEMPLATE *tmpl)
+        opaque_attr->pValue = (CK_BYTE *) opaque_attr + sizeof(CK_ATTRIBUTE);
+        memcpy(opaque_attr->pValue, aes_key, token_keysize);
+        rc = template_update_attribute(tmpl, opaque_attr);
+        if (rc != CKR_OK) {
+            TRACE_ERROR("template_update_attribute failed\n");
+            free(opaque_attr);
+            goto err;
+        }
+    } else {
+        if (token_keysize != key_size) {
+            TRACE_ERROR("Invalid key size: %lu\n", token_keysize);
+            rc = CKR_FUNCTION_FAILED;
+            goto err;
+        }
+    }
+
+    value_attr = (CK_ATTRIBUTE *) malloc(sizeof(CK_ATTRIBUTE) + key_size);
+    key_type_attr =
+        (CK_ATTRIBUTE *) malloc(sizeof(CK_ATTRIBUTE) + sizeof(CK_KEY_TYPE));
+    class_attr =
+        (CK_ATTRIBUTE *) malloc(sizeof(CK_ATTRIBUTE) + sizeof(CK_OBJECT_CLASS));
+    local_attr =
+        (CK_ATTRIBUTE *) malloc(sizeof(CK_ATTRIBUTE) + sizeof(CK_BBOOL));
+
+    if (!value_attr || !key_type_attr || !class_attr || !local_attr) {
+        TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+        rc = CKR_HOST_MEMORY;
+        goto err;
+    }
+
+    value_attr->type = CKA_VALUE;
+    value_attr->ulValueLen = key_size;
+    value_attr->pValue = (CK_BYTE *) value_attr + sizeof(CK_ATTRIBUTE);
+    if (is_opaque)
+        memset(value_attr->pValue, 0, key_size);
+    else
+        memcpy(value_attr->pValue, aes_key, key_size);
+    free(aes_key);
+    aes_key = NULL;
+
+    key_type_attr->type = CKA_KEY_TYPE;
+    key_type_attr->ulValueLen = sizeof(CK_KEY_TYPE);
+    key_type_attr->pValue = (CK_BYTE *) key_type_attr + sizeof(CK_ATTRIBUTE);
+    *(CK_KEY_TYPE *) key_type_attr->pValue = CKK_AES_XTS;
+
+    class_attr->type = CKA_CLASS;
+    class_attr->ulValueLen = sizeof(CK_OBJECT_CLASS);
+    class_attr->pValue = (CK_BYTE *) class_attr + sizeof(CK_ATTRIBUTE);
+    *(CK_OBJECT_CLASS *) class_attr->pValue = CKO_SECRET_KEY;
+
+    local_attr->type = CKA_LOCAL;
+    local_attr->ulValueLen = sizeof(CK_BBOOL);
+    local_attr->pValue = (CK_BYTE *) local_attr + sizeof(CK_ATTRIBUTE);
+    *(CK_BBOOL *) local_attr->pValue = TRUE;
+
+    rc = template_update_attribute(tmpl, value_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed\n");
+        goto err;
+    }
+    value_attr = NULL;
+    rc = template_update_attribute(tmpl, key_type_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed\n");
+        goto err;
+    }
+    key_type_attr = NULL;
+    rc = template_update_attribute(tmpl, class_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed\n");
+        goto err;
+    }
+    class_attr = NULL;
+    rc = template_update_attribute(tmpl, local_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("template_update_attribute failed\n");
+        goto err;
+    }
+    local_attr = NULL;
+
+    return CKR_OK;
+
+err:
+    if (aes_key)
+        free(aes_key);
```

Comment:
The free routine can handle NULL args, so it's not necessary to do the checks. On the other hand they don't hurt ...